File: vscode-extension/src/review/housekeeping.ts
/**
 * Acciones de housekeeping y su mapeo a la lista cerrada de args
 * (specs/006-superficie-panel-completa/contracts/cli-invocation.md).
 * Sin vscode: puro y testeable.
 */

import {sourceOf} from "../cli/porcelain";
import type {FixesState} from "../cli/porcelain";
import type {ReviewState} from "./state";

export type HousekeepingKind =
    | "clean-one"
    | "clean-keep-fixes"
    | "clean-fixes-only"
    | "clean-fixes-only-all"
    | "clean-all"
    | "forget-saved-one"
    | "forget-saved-all"
    | "forget-delta-one"
    | "forget-delta-all"
    | "forget-delta-stale";

export interface HousekeepingAction {
    kind: HousekeepingKind;
    /** Source name (`feature/x`), nunca `review-saved/x`. Requerido en `*-one` y `clean-keep-fixes`. */
    source?: string;
    /**
     * Solo `clean-keep-fixes`: si el finish fue con `--onto-source`. Define el
     * destino que nombra el confirm (rama del PR vs `review-fixes/<src>`).
     * Ausente se trata como no-onto (destino por defecto).
     */
    onto?: boolean;
    /**
     * Sólo `clean-fixes-only`: lo que la CLI reportó sobre esa rama, para que
     * la confirmación diga cuánto trabajo cuesta. No se deriva acá — quien
     * puede preguntarle a git es la CLI. Ausente se trata como `unknown`.
     */
    fixesState?: FixesState;
    /**
     * Sólo `clean-fixes-only`: si `review/<src>` sigue existiendo. Cambia la
     * copy y nada más: el argv lleva `--fixes-only` siempre, porque un dato que
     * se relee en cada refresco no puede decidir qué ramas borra un comando.
     */
    session?: boolean;
}

/**
 * Lo que la confirmación de `clean-fixes-only` dice sobre el costo. Una oración
 * por estado y ninguna se pliega en otra: "no committed" no es "seguro porque
 * ya está integrada", y "unknown" no es "no integrada".
 */
function fixesCostSentence(state: FixesState | undefined): string {
    switch (state) {
        case "empty":
            return "Nothing was ever committed on it, so no work of yours is lost.";
        case "merged":
            return "Its commits are already in the base branch.";
        case "unmerged":
            return "It has commits the base branch does not have — deleting it loses them.";
        default:
            return "There is no base branch configured, so git cannot tell whether its commits are integrated.";
    }
}

/** Prefijos de rama de review → source (mismo criterio que `sourceOf` en porcelain). */
export function sourceFromReviewName(name: string): string {
    for (const prefix of ["review-saved/", "review/", "review-fixes/"]) {
        if (name.startsWith(prefix)) {
            return name.slice(prefix.length);
        }
    }
    return name;
}

/**
 * Cierre `pending` a limpiar/deshacer desde el panel. No depende de HEAD: el
 * pending es del repo (`list --porcelain`), igual que `finish --abort`.
 *
 * Ausente fuera de `finish-pending`, o si el inventario no trae fila pending
 * (no debería ocurrir cuando la situación es esa).
 */
export function pendingFinishInfo(
    state: ReviewState
): {source: string; onto: boolean} | undefined {
    if (state.situation !== "finish-pending") {
        return undefined;
    }
    const pending = state.branches.find((branch) => branch.finish?.state === "pending");
    if (!pending?.finish) {
        return undefined;
    }
    return {source: sourceOf(pending), onto: pending.finish.onto};
}

/** Source del pending de finish, o `undefined`. Ver `pendingFinishInfo`. */
export function pendingFinishSource(state: ReviewState): string | undefined {
    return pendingFinishInfo(state)?.source;
}

export function verbForHousekeeping(action: HousekeepingAction): "clean" | "forget" {
    return action.kind.startsWith("clean") ? "clean" : "forget";
}

/**
 * Args del verbo (sin el verbo). Lanza si falta `source` en un kind que lo
 * exige (`*-one`, `clean-keep-fixes`).
 */
export function argsForHousekeeping(action: HousekeepingAction): string[] {
    switch (action.kind) {
        case "clean-one":
            if (!action.source) {
                throw new Error("clean-one requires source");
            }
            return [action.source];
        case "clean-keep-fixes":
            if (!action.source) {
                throw new Error("clean-keep-fixes requires source");
            }
            // Flag before the branch name — same shape as the CLI usage line.
            return ["--keep-fixes", action.source];
        case "clean-fixes-only":
            if (!action.source) {
                throw new Error("clean-fixes-only requires source");
            }
            // Flag before the branch name, same shape as --keep-fixes.
            return ["--fixes-only", action.source];
        case "clean-fixes-only-all":
            // No branch: --fixes-only alone only ever touches review-fixes/*
            // (clean's own scoping, see bin/git-review-verbs/clean), so this
            // never reaches a live review/* session the way a bare clean-all does.
            return ["--fixes-only"];
        case "clean-all":
            return [];
        case "forget-saved-one":
            if (!action.source) {
                throw new Error("forget-saved-one requires source");
            }
            return ["--saved", action.source];
        case "forget-saved-all":
            return ["--saved", "--all"];
        case "forget-delta-one":
            if (!action.source) {
                throw new Error("forget-delta-one requires source");
            }
            return ["--delta", action.source];
        case "forget-delta-all":
            return ["--delta", "--all"];
        case "forget-delta-stale":
            return ["--delta", "--stale"];
        default: {
            const _exhaustive: never = action.kind;
            throw new Error(`unknown housekeeping kind: ${_exhaustive}`);
        }
    }
}

/** `forget --delta --stale` toca la red (fetch). */
export function housekeepingNeedsNetwork(action: HousekeepingAction): boolean {
    return action.kind === "forget-delta-stale";
}

export interface ConfirmCopy {
    title: string;
    detail: string;
    button: string;
}

export function confirmCopyFor(action: HousekeepingAction): ConfirmCopy {
    const src = action.source ?? "";
    switch (action.kind) {
        case "clean-one":
            return {
                title: `Delete the leftovers from reviewing ${src}?`,
                detail: `This removes the review branch and any edits you extracted from it. Anything you already committed elsewhere stays. It cannot be undone.`,
                button: "Delete",
            };
        case "clean-keep-fixes": {
            // onto → edits en la rama del PR; sin onto → review-fixes/<src>.
            // El clean no toca el working tree: solo tira el undo review/*.
            //
            // Lo que se CONSERVA va primero, y no es un adorno: este diálogo
            // sale del botón que cierra el ciclo, y la única duda que frena a
            // alguien ahí es si el clean se lleva sus ediciones. La respuesta
            // es que no, y decirla antes de nombrar lo que sí se pierde es la
            // diferencia entre leer el cartel y apretar a ciegas.
            const destination = action.onto ? src : `review-fixes/${src}`;
            return {
                title: `Done with the review of ${src}?`,
                detail: `Your edits stay on ${destination} — commit and push them from Source Control. What goes away is being able to undo this finish.`,
                button: "Done",
            };
        }
        case "clean-fixes-only": {
            // La sesión se nombra sólo cuando existe: prometer que se deja algo
            // que no está es ruido, y el argv es el mismo en los dos casos.
            const session = action.session
                ? " You can still undo the finish afterwards."
                : "";
            return {
                title: `Delete the edits you extracted from ${src}?`,
                detail: `${fixesCostSentence(action.fixesState)}${session} It cannot be undone.`,
                button: "Delete",
            };
        }
        case "clean-fixes-only-all":
            return {
                title: "Delete every branch of extracted edits?",
                detail: "They hold edits you made while reviewing and never committed anywhere else. Nothing you are reviewing right now is touched. It cannot be undone.",
                button: "Delete all",
            };
        case "clean-all":
            return {
                title: "Delete all review leftovers?",
                detail: "This removes every review branch and every branch of extracted edits that you are not currently on. Paused reviews and your last review points are left alone. It cannot be undone.",
                button: "Delete all",
            };
        case "forget-saved-one":
            return {
                title: `Delete the paused review of ${src}?`,
                detail: `This throws away the edits you had saved with it. It cannot be undone.`,
                button: "Delete",
            };
        case "forget-saved-all":
            return {
                title: "Delete every paused review?",
                detail: "This throws away the edits saved with each of them. It cannot be undone.",
                button: "Delete all",
            };
        // Los tres de --delta dicen la CONSECUENCIA y no la operación, y la
        // dicen con la etiqueta que el asistente usa para el rango ("only what
        // is new"): quien vaya a apretar esto lo eligió alguna vez ahí, y es el
        // único lugar donde ese dato se nota. "Removes the last-reviewed tip"
        // describía un ref que ninguna superficie del producto nombra.
        case "forget-delta-one":
            return {
                title: `Forget where you got to on ${src}?`,
                detail: "Next time you review this branch, \"only what is new\" will have no starting point, so you will be offered the full range instead.",
                button: "Forget",
            };
        case "forget-delta-all":
            return {
                title: "Forget where you got to on every branch?",
                detail: "Next time you review any of them, \"only what is new\" will have no starting point, so you will be offered the full range instead.",
                button: "Forget all",
            };
        case "forget-delta-stale":
            return {
                title: "Forget the branches that are gone?",
                detail: "This clears where you got to on branches that no longer exist. It checks the remote first, so it may take a moment.",
                button: "Forget",
            };
        default: {
            const _exhaustive: never = action.kind;
            throw new Error(`unknown housekeeping kind: ${_exhaustive}`);
        }
    }
}

File: intellij-plugin/src/main/kotlin/com/ezevillo/gitreview/domain/Housekeeping.kt
package com.ezevillo.gitreview.domain

enum class HousekeepingKind {
    CLEAN_ONE,
    CLEAN_KEEP_FIXES,
    CLEAN_ALL,
    FORGET_SAVED_ONE,
    FORGET_SAVED_ALL,
    FORGET_DELTA_ONE,
    FORGET_DELTA_ALL,
    FORGET_DELTA_STALE,
}

data class HousekeepingAction(
    val kind: HousekeepingKind,
    val source: String? = null,
    val onto: Boolean? = null,
)

data class ConfirmCopy(
    val title: String,
    val detail: String,
    val button: String,
)

fun sourceFromReviewName(name: String): String {
    for (prefix in listOf("review-saved/", "review/", "review-fixes/")) {
        if (name.startsWith(prefix)) return name.removePrefix(prefix)
    }
    return name
}

fun pendingFinishInfo(state: ReviewState): Pair<String, Boolean>? {
    if (state.situation != Situation.FINISH_PENDING) return null
    val pending = state.branches.find { it.finish?.state == "pending" } ?: return null
    val finish = pending.finish ?: return null
    return sourceOf(pending) to finish.onto
}

fun pendingFinishSource(state: ReviewState): String? = pendingFinishInfo(state)?.first

fun verbForHousekeeping(action: HousekeepingAction): String =
    if (action.kind.name.startsWith("CLEAN")) "clean" else "forget"

fun argsForHousekeeping(action: HousekeepingAction): List<String> = when (action.kind) {
    HousekeepingKind.CLEAN_ONE -> {
        require(!action.source.isNullOrEmpty()) { "clean-one requires source" }
        listOf(action.source)
    }
    HousekeepingKind.CLEAN_KEEP_FIXES -> {
        require(!action.source.isNullOrEmpty()) { "clean-keep-fixes requires source" }
        listOf("--keep-fixes", action.source)
    }
    HousekeepingKind.CLEAN_ALL -> emptyList()
    HousekeepingKind.FORGET_SAVED_ONE -> {
        require(!action.source.isNullOrEmpty()) { "forget-saved-one requires source" }
        listOf("--saved", action.source)
    }
    HousekeepingKind.FORGET_SAVED_ALL -> listOf("--saved", "--all")
    HousekeepingKind.FORGET_DELTA_ONE -> {
        require(!action.source.isNullOrEmpty()) { "forget-delta-one requires source" }
        listOf("--delta", action.source)
    }
    HousekeepingKind.FORGET_DELTA_ALL -> listOf("--delta", "--all")
    HousekeepingKind.FORGET_DELTA_STALE -> listOf("--delta", "--stale")
}

fun housekeepingNeedsNetwork(action: HousekeepingAction): Boolean =
    action.kind == HousekeepingKind.FORGET_DELTA_STALE

fun confirmCopyFor(action: HousekeepingAction): ConfirmCopy {
    val src = action.source.orEmpty()
    return when (action.kind) {
        HousekeepingKind.CLEAN_ONE -> ConfirmCopy(
            title = "Clean leftover review branches for $src?",
            detail = "Deletes review/$src and review-fixes/$src (and banked edit refs) if they exist and are not checked out. Does not touch delta markers.",
            button = "Clean",
        )
        HousekeepingKind.CLEAN_KEEP_FIXES -> {
            val destination = if (action.onto == true) src else "review-fixes/$src"
            ConfirmCopy(
                title = "Drop the finish undo for $src?",
                detail = "Runs git review clean --keep-fixes $src: deletes review/$src and the finish undo point so the pending finish goes away. Your staged edits stay on $destination; delta markers are left alone. Remember to commit and push them from Source Control.",
                button = "Clean",
            )
        }
        HousekeepingKind.CLEAN_ALL -> ConfirmCopy(
            title = "Clean all leftover review branches?",
            detail = "Deletes every review/* and review-fixes/* branch that is not currently checked out, plus orphaned edit/undo refs. Does not touch delta markers or saved reviews.",
            button = "Clean All",
        )
        HousekeepingKind.FORGET_SAVED_ONE -> ConfirmCopy(
            title = "Discard the saved review of $src?",
            detail = "Deletes review-saved/$src, its banked edits and metadata, and rolls back the delta marker it left.",
            button = "Discard",
        )
        HousekeepingKind.FORGET_SAVED_ALL -> ConfirmCopy(
            title = "Discard every saved review?",
            detail = "Deletes all review-saved/* branches, their banked edits and metadata, and rolls back their delta markers.",
            button = "Discard All Saved",
        )
        HousekeepingKind.FORGET_DELTA_ONE -> ConfirmCopy(
            title = "Forget the delta marker for $src?",
            detail = "Removes the last-reviewed tip used by git review start --delta for this branch (remote and local markers).",
            button = "Forget Marker",
        )
        HousekeepingKind.FORGET_DELTA_ALL -> ConfirmCopy(
            title = "Forget every delta marker?",
            detail = "Removes all last-reviewed tips used by git review start --delta.",
            button = "Forget All Markers",
        )
        HousekeepingKind.FORGET_DELTA_STALE -> ConfirmCopy(
            title = "Forget stale delta markers?",
            detail = "Fetches from the remote (when needed) and removes markers whose branch no longer exists.",
            button = "Forget Stale",
        )
    }
}
